### PR TITLE
docs: harvest review lessons from this week's PRs into internal docs

### PR DESCRIPTION
**Pull request by saltas888**
## Why

Reviewers repeatedly had to spell out the same conventions on this week's PRs, because each lesson lived in a review thread and died there. This is a `harvesting-review` run over PRs merged 2026-07-21 → 2026-07-24 (the window since the last such run, [#9991](https://github.com/opsmill/infrahub/pull/9991)), turning the feedback that generalizes into durable internal-doc conventions so a future author (human or agent) follows the convention the first time.

Each candidate lesson was reconstructed against the before/after code, verified against the current codebase, and checked against what's already documented before landing — per `.claude/skills/harvesting-review/`.

## Source PRs

Ten PRs were audited (research delegated to one subagent per PR); the ones that yielded a durable lesson:

- **#9948** (bilalabbad, column-header sort & filter menu) — pa-lem: `PopoverDialog`'s `useDialog` steals focus from a nested `defaultOpen` combobox and closes it; bilalabbad: sort change without pagination reset causes a page-stitching bug on large datasets (fix deferred to a follow-up PR).
- **#10028** (polmichel, regeneration safe-fallback escalation) — cubic-dev-ai: a spec doc's scenario table cited test functions that don't exist anywhere in the repo yet.
- **#9996** (ajtmccarty, memory-efficient rollback query) — cubic-dev-ai: cleanup driven by ids collected in an earlier phase is lost if the process crashes between phases; the fix is already documented for the merge-recovery domain but not surfaced in the general migrations guide.
- **#9994** (ajtmccarty, node-scoped uniqueness constraint queries) — polmichel: UUID/liveness-check queries need an explicit test for the duplicate-UUID-from-migration case; the invariant is documented but the testing guide never points authors at it.

Also scanned with no new lessons to harvest (already covered, PR-local, in-flight elsewhere, universal hygiene, or no substantive review): #10027, #10025, #9990, #9791, #9995, #10000.

**Not duplicated:** #10000 surfaced several real lessons (N+1/set-based Cypher writes, exception-hierarchy verification before narrowing, an unreleased-feature changelog carve-out, ticket IDs in internal docs, dynamic Cypher labels) — all of these are already being fixed in the open PR [#10005](https://github.com/opsmill/infrahub/pull/10005) (`dev/guidelines/git-workflow.md`, `checklist.md`, `python.md`, `changelog.md`, `documentation.md`, the `neo4j-cypher-guide` skill). Left alone rather than duplicated or conflicted, matching how #9991 excluded lessons already in-flight on other open PRs.

## What changed

**New rules:**

- `dev/knowledge/frontend/design-system.md`: `PopoverDialog` wraps react-aria's `useDialog`, whose focus-on-mount steals focus from a nested `defaultOpen` combobox/dropdown portal and closes it — documents the carve-out to use plain `Popover` for that content instead.
- `dev/guidelines/frontend/page-architecture.md`: a new rule that an offset-paginated list with user-driven sorting must reset to page 1 on sort change, plus an anti-pattern table entry citing the PR.
- `dev/guidelines/documentation.md`: before citing a specific test as verifying a scenario in a spec/status doc, confirm that test exists — label unwritten coverage as planned/unverified instead.

**Strengthened existing coverage (rule existed but wasn't landing — "not discoverable" per the harvesting-review diagnosis):**

- `dev/guides/backend/creating-migrations.md`: cross-links the candidate-ids-across-phases pitfall already written up in `merge-failure-recovery.md` — the general migration guide never pointed authors at it.
- `dev/knowledge/backend/testing.md`: cross-links the duplicate-UUID invariant already documented in `database-schema.md` so test authors are prompted to cover it instead of relying on a reviewer to ask.

Docs-only — no code, no generated files, no schema.

## How to review

None of the five edits are `.agents/rules/*` (no per-turn token cost added) — all land in `dev/guidelines/` or `dev/knowledge/`, loaded only via each area's router. The two "strengthen" edits are single-sentence cross-links to existing content, not new rules.

## How to test

```bash
npx -y markdownlint-cli2 dev/guidelines/documentation.md dev/guidelines/frontend/page-architecture.md dev/guides/backend/creating-migrations.md dev/knowledge/backend/testing.md dev/knowledge/frontend/design-system.md
```

0 issues.

## Impact & rollout

- **Backward compatibility:** docs-only, no runtime impact
- **Deployment notes:** safe to merge

## Checklist

- [ ] Tests added/updated (n/a — docs-only)
- [ ] Changelog entry added (n/a — internal docs)
- [ ] External docs updated (n/a)
- [x] Internal .md docs updated (internal knowledge and AI code tools knowledge)
- [x] I have reviewed AI generated content

🤖 Generated with [Claude Code](https://claude.ai/code)

https://claude.ai/code/session_01USfYPX72WZAoSeRjKHiUDR

---
_Generated by [Claude Code](https://claude.ai/code/session_01USfYPX72WZAoSeRjKHiUDR)_

<a href="https://cubic.dev/pr/opsmill/infrahub/pull/10032?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: dev/guidelines/frontend/page-architecture.md [hunk 1]
```diff
@@ -96,6 +96,10 @@ If the backend already filters, transforms, or defaults something, do not duplic
 
 If the client needs to *display* a server-side default, surface it via the API (extend a query or response field) rather than mirroring the constant.
 
+## Pagination must reset when sort changes
+
+An offset-paginated list that also supports user-driven sorting must reset its offset to page 1 whenever the active sort changes. Preserving the current offset across a sort change causes a page-stitching bug: at a deep offset, switching sort direction shows the tail of the new order immediately followed by rows from the old offset — two inconsistent windows stitched together. Reset-to-page-1-on-sort-change is the safe default; don't thread the previous offset through independently of sort state.
```
[the remark this answers is not in the document]
Comment (saltas888) [reply dev/guidelines/frontend/page-architecture.md]: Correct — verified `dev/specs/002-header-sort-menu/spec.md` FR-010 and the passing `table-column-header.test.tsx` test both intentionally preserve the offset today. Reworded in 6e856e3 to describe this as the reviewer-identified, deferred bug it actually is, and scoped the forward-looking guidance to new lists rather than asserting current code is non-compliant.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: dev/guides/backend/creating-migrations.md [hunk 2]
```diff
@@ -111,6 +111,8 @@ CALL (n) {
 
 `CALL ... IN TRANSACTIONS` requires the `MATCH` to be outside the subquery — move all matching up front.
 
+**Pitfall — don't carry candidate ids across phases in application memory.** A multi-phase migration (phase 1 computes candidate node/edge ids, phase 2 acts on them — e.g. deleting orphans or restoring metadata for what phase 1 touched) must not hold those candidates in a Python list to drive the later phase. A crash between phases loses that in-memory list, and a resumed run can no longer tell what still needs cleanup. Fold the follow-up phase into the same `CALL ... IN TRANSACTIONS` block as the phase that produces the candidates, re-deriving them by the same filter, so an interrupted-and-resumed migration is safe. See [Merge Failure Recovery](../../knowledge/backend/merge-failure-recovery.md) for a worked example of this fix.
```
[the remark this answers is not in the document]
Comment (saltas888) [reply dev/guides/backend/creating-migrations.md]: Right, that wording could be read as "merge every phase into one statement," which contradicts the recovery doc's own warning against chaining phases (transaction-memory exhaustion). Reworded in 6e856e3: cleanup stays paired with the pass that produces its candidates, but distinct logical phases can remain separate passes when the planner/memory requires it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: dev/guidelines/frontend/page-architecture.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (saltas888) [reply dev/guidelines/frontend/page-architecture.md]: Confirmed `DEFAULT_OFFSET = 0` in `usePagination.ts`. Fixed in 6e856e3 to use "offset" terminology throughout instead of "page 1".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: dev/knowledge/frontend/design-system.md [hunk 3]
```diff
@@ -75,6 +75,8 @@ If a component needs a slot the package doesn't expose yet, **extend the package
 
 **Popover exists in both places.** `shared/components/ui/popover.tsx` is Radix-based (`@radix-ui/react-popover`), adds `PopoverAnchor` and `PopoverTabs*` subcomponents, and marks its content with `data-react-aria-top-layer` so clicking inside it does not count as an outside interaction for an enclosing react-aria overlay such as the package `Sheet`. `@infrahub/ui` exports a separate react-aria `Popover` / `PopoverDialog` / `PopoverTrigger`. Both are actively used in the app (roughly 23 vs 26 importing files); the dismiss guard itself lives in `@infrahub/ui` (`Sheet` + `useDismissGuard`), not in either popover.
 
+**`PopoverDialog` gotcha: don't wrap content that auto-opens a nested combobox/dropdown.** `PopoverDialog` wraps react-aria's `Dialog`, which invokes `useDialog` and focuses itself on mount. If the popover's content renders a `defaultOpen` combobox or dropdown in its own portal (e.g. an enum/relationship filter field), `useDialog`'s focus-on-mount steals focus from that nested portal and closes it immediately. For content shaped like that, skip `PopoverDialog` and render the plain `Popover` instead — accepting the reduced dialog semantics for that one case.
```
[the remark this answers is not in the document]
Comment (saltas888) [reply dev/knowledge/frontend/design-system.md]: Confirmed — `ComboboxContent` (used by `Enum`/`Dropdown`, which back the filter fields) sets `portal={false}`. Dropped the "in its own portal" claim in 6e856e3; the focus-steal mechanism doesn't depend on portal usage anyway.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: dev/guidelines/frontend/page-architecture.md
Comment (ajtmccarty) [inline dev/guidelines/frontend/page-architecture.md:1]: do we want PRs and spec references in here? actually asking, I'm not sure

**Review by ajtmccarty — APPROVED** [review]

File: dev/guidelines/frontend/page-architecture.md
Comment (polmichel) [reply dev/guidelines/frontend/page-architecture.md:1]: In my opinion, I'd rather remove these references, since they can go stale. The code should only contain comments that describe how the system works, and so are the internal guidelines. Major architectural decisions that are hard to grasp from the code should be tracked in ADRs, and even those should stay as independent from the spec as possible and explain the underlying problem instead.

This is admittedly subjective, so I'm obviously open to discussion.

Comment (saltas888) [reply dev/guidelines/frontend/page-architecture.md:1]: I agree with you guys, good point. We should be archiving them and collecting feedback

Comment (saltas888) [reply dev/guidelines/frontend/page-architecture.md:1]: I updated the dev/guidelines/documentation.md so it knows how to document better next time
